answers.md

```diff
@@ -0,0 +1,50 @@
+# Answers
+## HPC good citizenship
+1. No. `-pe openmp 64` only tells the grid engine to allocate 64 cores. But the real number of processors used are determined by the programs running on it.
+
+
+2. Review the scripts before submitting to the grid engine. Softwares written by others usually have documentations. If not, try running the softwares with a smaller dataset as a test.
+
+
+3. Use following command
+
+```
+/usr/bin/time -f "maximum resident set size = %M, elapsed real time = %E, cpu time = %S, user time = %U" command
+```
+The idea is from this [page](https://unix.stackexchange.com/questions/375889/unix-command-to-tell-how-much-ram-was-used-during-program-runtime).
+
+
+4. For the maximum resident set size, Kbytes. For the elapsed real time, [hours:]minutes:seconds. For cpu time and user time, seconds.
+
+
+5. Use following commands
+```
+# Checking that a file exists
+[ -e filename ]
+# Checking that a file exists and is not empty
+[ -s filename ]
+```
+
+
+6. Use following command
+```
+[ -e filename ] || run-a-job
+```
+
+
+
+## Trickier questions
+7. No. macOS (Darwin) is one kind of UNIX-like OS. And the "time" is not the same as that on Linux, which is another branch of UNIX-like OS.
+
+
+8. Specify the RAM in the script to submit the job by adding `-l h_vmem=size`.
+```
+#!/bin/bash
+# Modified from the course material for EE282 by Edwin
+#$ -N jobname
+#$ -q nodenames
+#$ -l h_vmem=size=24G
```

Comment:
Hi, this just check for virtual memory and not resident memory. the correct option would be -l mem_free=24GB
I would like to add some caveats.

By default and as a common courtesy it is best to take the total amount of RAM on a system, subtract a small overhead (our admins have mostly taken care of this so we don't really need to worry about this value much) and then divide it by the total number of cores. if your job will be using 256GB of RAM on a 512GB RAM node it is best to also use the same portion of cores for your job, even though your job will only run on 1 core. Technically, our grid should have the option to limit the amount of maximum memory requested based on number of cores requested. i.e on a 64 core 512GB node the user should allocate a maximum of 8GB per core. if the single core job were to use 16GB then it would require that 2 cores be allocated thereby avoiding issues caused by the over consumption of RAM by a job.